### PR TITLE
Fix cancellation leaks

### PR DESCRIPTION
**Pull request by sharwell**
Failing to dispose of a `CancellationTokenSource` created with `CreateLinkedTokenSource` will result in the underlying sources never clearing their callbacks which point to the linked source. If one or more of the underlying tokens is associated with a long-lived token source, this will leak a `CancellationTokenSource` and a `CancellationCallbackInfo` (used by `CancellationTokenRegistration`).

⚠️ `CancellationTokenSource.Cancel` is not an alternative to calling `Dispose()`. The token source must be explicitly disposed even after cancellation in order to prevent leaks.

Failing to dispose of a `CancellationTokenRegistration` results in both the above and adds a strong reference to a captured execution context, which includes strong references to the current value of an `AsyncLocal<T>`. This can be an especially large leak when the latter references a `JoinableTask<T>` which is part of [`JoinableTaskContext.joinableOperation`](https://github.com/Microsoft/vs-threading/blob/955829f236251e0097a469a2aed894c8f81c8bdb/src/Microsoft.VisualStudio.Threading/JoinableTaskContext.cs#L89).

This pull requests ensures that these types are disposed where they are created within the Roslyn code base.

<details><summary>Ask Mode not complete</summary><p>

**Customer scenario**

What does the customer do to get into this situation, and why do we think this
is common enough to address for this release.  (Granted, sometimes this will be
obvious "Open project, VS crashes" but in general, I need to understand how
common a scenario is)

**Bugs this fixes:**

(either VSO or GitHub links)

**Workarounds, if any**

Also, why we think they are insufficient for RC vs. RC2, RC3, or RTW

**Risk**

This is generally a measure our how central the affected code is to adjacent
scenarios and thus how likely your fix is to destabilize a broader area of code

**Performance impact**

(with a brief justification for that assessment (e.g. "Low perf impact because no extra allocations/no complexity changes" vs. "Low")

**Is this a regression from a previous update?**

**Root cause analysis:**

How did we miss it?  What tests are we adding to guard against it in the future?

**How was the bug found?**

(E.g. customer reported it vs. ad hoc testing)

</p></details>

### CODE REVIEW

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs [hunk 1]
```diff
@@ -192,6 +192,8 @@ private static Task<MSB.Execution.BuildResult> BuildAsync(MSB.Execution.BuildMan
                         taskSource.TrySetCanceled();
                     }
                 });
+
+                taskSource.Task.SafeContinueWith(_ => registration.Dispose(), CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs]: Just to verify, but registration is safe to double dispose?

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs]: i'm not a huge fan of hte SafeContinueWith approach.  Can we instead just call an async method that awaits taskSource, then calls registrtion.Dispose in a finally?

**Comment by CyrusNajmabadi** [comment]
?w=1 makes this much easier to review :)

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs (hunk 1 above)
Comment (sharwell) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs]: > Just to verify, but registration is safe to double dispose?

It should be, but either way it seems there's no longer any need for the second call so I can remove it.

> i'm not a huge fan of hte SafeContinueWith approach.

It reduces the overhead of the cleanup by allowing the use of `ExecuteSynchronously`.

**Comment by CyrusNajmabadi** [comment]
Doesn't the TPL allow for synchronous execution on awaits?

**Comment by sharwell** [comment]
@CyrusNajmabadi Updated per feedback. Commit ed84156 switches from `SafeContinueWith` to a cleanup function. The function seems general purpose for the task, but I didn't know where else to put it so I left it as a local function.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs [hunk 2]
```diff
@@ -850,6 +851,8 @@ private int RunCore(TextWriter consoleOutput, ErrorLogger errorLogger, Cancellat
                         ReportAnalyzerExecutionTime(consoleOutput, analyzerDriver, Culture, compilation.Options.ConcurrentBuild);
                     }
                 }
+
+                analyzerDriver?.Dispose();
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: Need @dotnet/roslyn-compiler review.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 3]
```diff
@@ -1369,9 +1369,18 @@ private static bool IsCompilerAnalyzer(DiagnosticAnalyzer analyzer)
...
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposing)
+            {
+                this.CompilationEventQueue?.TryComplete();
+                this.DiagnosticQueue?.TryComplete();
+                _queueRegistration.Dispose();
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: need @dotnet/roslyn-analysis review.

File: src/Interactive/Features/Interactive/Core/InteractiveHost.LazyRemoteService.cs [hunk 4]
```diff
@@ -62,20 +62,23 @@ public void Dispose(bool joinThreads)
                     }
 
                     bool initializing = true;
-                    cancellationToken.Register(() =>
-                    {
-                        if (initializing)
+                    RemoteExecutionResult initializationResult;
+                    using (cancellationToken.Register(
+                        () =>
```
Comment (CyrusNajmabadi) [inline src/Interactive/Features/Interactive/Core/InteractiveHost.LazyRemoteService.cs]: this seems... odd.  You're registering a cancellation callback, but you're immedaitely disposing?  I don't really understand what teh intent is here.  Can you comment the code?

File: src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs [hunk 5]
```diff
@@ -185,13 +186,35 @@ private static Task<MSB.Execution.BuildResult> BuildAsync(MSB.Execution.BuildMan
...
                     finally
                     {
                         taskSource.TrySetCanceled();
                     }
                 });
+
+                // Dispose of the registration as soon as we no longer need it
+                var _ = DisposeWhenCompleteAsync(taskSource.Task, registration);
+
+                async Task DisposeWhenCompleteAsync<T>(Task task, T disposable)
```
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs]: I wouldn't mind this as a general extension on Task that goes with our other task extensions.  It seems out of place here to be generic, but only called with one arg in one place.

Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Desktop/Workspace/MSBuild/ProjectFile/ProjectFile.cs]: can you move the local function to the bottom of the method (or as a sibling method).  I find it *very* hard to read methods that have a local function randomly interspersed with the code.

**Comment by CyrusNajmabadi** [comment]
Done with review pass.  There's one part i don't understand and in need info on.  We also need compiler/analyzer review as this touches those layers.

File: src/Interactive/Features/Interactive/Core/InteractiveHost.LazyRemoteService.cs (hunk 4 above)
Comment (sharwell) [inline src/Interactive/Features/Interactive/Core/InteractiveHost.LazyRemoteService.cs]: The callback is a NOP if `initializing` is false. The using statement ensures the object is disposed at the point where it would no longer have any work to do. I'm not sure what the original intent was, since I was just preserving the original semantics. Do you have a suggested comment to add?

Comment (CyrusNajmabadi) [inline src/Interactive/Features/Interactive/Core/InteractiveHost.LazyRemoteService.cs]: I think we need @tmat to help us out here.

Comment (VSadov) [inline src/Interactive/Features/Interactive/Core/InteractiveHost.LazyRemoteService.cs]: I am not sure about the intent, but it seems you are deregistering the callback earlier now.

---
In reply to: [121204270](https://github.com/dotnet/roslyn/pull/20111#discussion_r121204270) [](ancestors = 121204270)

File: src/Interactive/Features/Interactive/Core/InteractiveHost.LazyRemoteService.cs [hunk 6]
```diff
@@ -62,20 +62,23 @@ public void Dispose(bool joinThreads)
...
-                        service.InitializeContextAsync(operation, Options.InitializationFile, isRestarting: InstanceId > 1);
-                    }).ConfigureAwait(false);
+                        // try to execute initialization script:
+                        initializationResult = await Async<RemoteExecutionResult>(remoteService, (service, operation) =>
+                        {
+                            service.InitializeContextAsync(operation, Options.InitializationFile, isRestarting: InstanceId > 1);
+                        }).ConfigureAwait(false);
+                    }
 
                     initializing = false;
```
Comment (CyrusNajmabadi) [inline src/Interactive/Features/Interactive/Core/InteractiveHost.LazyRemoteService.cs]: Ohhhhhhhhh.  I see now.  Man, this was hard to parse.  I thought the part of the code in the using was actually in teh lambda.  

Can we pull the 'using' argument into a local variable to make it clearer what's going on?  Having a complex lambda in the expression part of a using-block is super hard to read.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 7]
```diff
@@ -1369,9 +1369,18 @@ private static bool IsCompilerAnalyzer(DiagnosticAnalyzer analyzer)
...
         public void Dispose()
         {
-            this.CompilationEventQueue?.TryComplete();
-            this.DiagnosticQueue?.TryComplete();
-            _queueRegistration.Dispose();
+            Dispose(true);
+            GC.SuppressFinalize(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: >protected virtual void Dispose(bool disposing) [](start = 8, length = 46)

I am not sure why this method is added. It looks like no one overrides it, there is only one call site and we always pass true.

File: src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs (hunk 2 above)
Comment (VSadov) [inline src/Compilers/Core/Portable/CommandLine/CommonCompiler.cs]: looks ok to me

---
In reply to: [121199682](https://github.com/dotnet/roslyn/pull/20111#discussion_r121199682) [](ancestors = 121199682)

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs (hunk 7 above)
Comment (sharwell) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: Since the enclosing type implements `IDisposable` and is not sealed, I fixed the implementation to adhere to requirements of the `IDisposable` interface. I believe this change can be moved to a separate PR if needed.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: I do not believe there is such a requirement. There is a recommendation about implementing IDisposable for classes that have finalizers/destructors, but that requires both implementations to be coordinated with each other. Simply adding one part doesn't feel useful to me (even in a separate PR), either we should complete the pattern, or remove the new method.

File: src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs [hunk 8]
```diff
@@ -1369,9 +1369,18 @@ private static bool IsCompilerAnalyzer(DiagnosticAnalyzer analyzer)
 
         public void Dispose()
         {
-            this.CompilationEventQueue?.TryComplete();
-            this.DiagnosticQueue?.TryComplete();
-            _queueRegistration.Dispose();
+            Dispose(true);
+            GC.SuppressFinalize(this);
```
Comment (VSadov) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: >SuppressFinalize [](start = 15, length = 16)

It does not look like this class is finalizable or holds any unmanaged resources.  
It seems strange to communicate with GC in an ordinary managed class.

Comment (sharwell) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: The implementation of `IDisposable.Dispose()` is sealed. Unsealed disposable types include a `GC.SuppressFinalize(this)` even if they do not directly own unmanaged resources to ensure correct behavior for a derived type which introduces a finalizer.

Comment (AlekseyTs) [inline src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs]: > The implementation of IDisposable.Dispose() is sealed. 

Implementation of an interface cannot be sealed, derived types can always re-implement any interface method. The class is internal, we control all concrete implementations of this class. Once a derived type adds a finalizer, we will deal with correct implementation of IDisposable, the finalizer will have to be written in a special way, nothing will just work because we added this code.

**Comment by sharwell** [comment]
The latest update squashed the changes related `CancellationTokenRegistration` cleanup. The specific changes were:

* Revert all changes to **AnalyzerDriver.cs**
* Move local function to end of containing method in **ProjectFile.cs**
* Simplify and document cancellation behavior in **InteractiveHost.LazyRemoteService.cs**

**Comment by sharwell** [comment]
@dotnet-bot retest this please

File: src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs [hunk 9]
```diff
@@ -87,8 +87,14 @@ private class QueueItem<TRequestType, TResponseType> : IQueueItem
...
             {
-                // Set the tcs state to cancelled if the token gets cancelled outside of our callback (for example the server shutting down).
-                cancellationToken.Register(() => _completionSource.TrySetCanceled(cancellationToken));
+                if (cancellationToken.CanBeCanceled)
+                {
+                    // Set the tcs state to cancelled if the token gets cancelled outside of our callback (for example the server shutting down).
+                    var registration = cancellationToken.Register(() => _completionSource.TrySetCanceled(cancellationToken));
+
+                    // Dispose of the registration as soon as we no longer need it
+                    _completionSource.Task.ContinueWith(_ => registration.Dispose(), CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
```
Comment (Therzok) [inline src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs:96]: Heh, I think you might want to use Unregister here, because it may cause deadlocks: https://source.dot.net/#System.Private.CoreLib/CancellationTokenRegistration.cs,f05a1300415bcfde,references

It's noted very obscurely in the docs [here](https://source.dot.net/#System.Private.CoreLib/CancellationTokenRegistration.cs,f05a1300415bcfde,references):

> If you call [Dispose](https://docs.microsoft.com/en-us/dotnet/api/system.threading.cancellationtokenregistration.dispose) while the callback is running, and you hold a lock that the callback is waiting on, your program can deadlock. After Dispose returns, you can free any resources required by the callback.

Comment (sharwell) [reply src/Features/LanguageServer/Protocol/Handler/RequestExecutionQueue.QueueItem.cs]: I pushed an update to address the straightforward cases in production code.

**Comment by AlekseyTs** [comment]
@sharwell What is the plan with this PR? It is pretty old, I mean by age. Is your intent to merge it soon?
